lib/dialect/include/rlc/dialect/ConstraintsAnalysis.hpp

```diff
@@ -0,0 +1,299 @@
+/*
+Copyright 2024 Matteo Cenzato
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+#pragma once
+
+#include <utility>
+#include <iostream>
+#include <limits>
+#include <optional>
+#include <string>
+
+#include "llvm/ADT/TypeSwitch.h"
+#include "mlir/Analysis/DataFlow/ConstantPropagationAnalysis.h"
+#include "mlir/Analysis/DataFlow/DeadCodeAnalysis.h"
+#include "mlir/Analysis/DataFlow/DenseAnalysis.h"
+#include "mlir/IR/BuiltinDialect.h"
+#include "mlir/Interfaces/ControlFlowInterfaces.h"
+#include "rlc/dialect/Operations.hpp"
+#include "rlc/dialect/Passes.hpp"
+#include "rlc/dialect/conversion/TypeConverter.h"
+#include "llvm/IR/ConstantRange.h"
+
+// Add the map key to the llvm scope
+// I have to admit that if this works the llms will take over the world someday
+// Ok it compiles
+// It seems to work. Wow
+namespace llvm{
+	template <>
+	struct llvm::DenseMapInfo<std::pair<mlir::Value,bool>>{
```

Comment:
we should make a struct out of this std::pair<mlir::Value, bool> and have the dense map being about it. The reason to do so is because otherwise if someone includes this file will get a different denseMapInfo for that pair alone than places that do not include the file